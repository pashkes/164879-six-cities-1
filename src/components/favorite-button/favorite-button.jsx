import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {withRouter} from "react-router-dom";
import {compose} from "recompose";

import {Operation} from "../../reducer/data/data";
import {getAuthorizationStatus} from "../../reducer/user/selectors";
import Constants from "./../../constants";

export class FavoriteButton extends PureComponent {
  constructor(props) {
    super(props);

    this._handleClick = this._handleClick.bind(this);
  }

  _handleClick() {
    const {
      isFavorite,
      onAddToFavorites,
      id,
      onRemoveFromFavorites,
      isAuthRequire,
      history,
    } = this.props;

    if (isAuthRequire) {
      history.push(Constants.LOGIN_PATH);
      return;
    }
    if (isFavorite) {
      onRemoveFromFavorites(id);
    } else {
      onAddToFavorites(id);
    }
  }

  render() {
    const {isFavorite, prefixClass, width = 18, height = 19} = this.props;

    return (
      <button
        onClick={this._handleClick}
        type="button"
        className={`${prefixClass}__bookmark-button ${isFavorite ? `place-card__bookmark-button--active` : ``} button`}
        aria-pressed={isFavorite ? `true` : `false`}
      >
        <svg className={`place-card__bookmark-icon`} aria-hidden="true" width={width} height={height}>
          <use xlinkHref="#icon-bookmark"/>
        </svg>
        <span className="visually-hidden">{isFavorite ? `In` : `To`} bookmarks</span>
      </button>
    );
  }
}

FavoriteButton.propTypes = {
  isFavorite: PropTypes.bool.isRequired,
  onAddToFavorites: PropTypes.func.isRequired,
  onRemoveFromFavorites: PropTypes.func.isRequired,
  id: PropTypes.number.isRequired,
  prefixClass: PropTypes.string.isRequired,
  width: PropTypes.string,
  height: PropTypes.string,
  isAuthRequire: PropTypes.bool.isRequired,
  history: PropTypes.object,
};

const mapStateToProps = (state) => ({
  isAuthRequire: getAuthorizationStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  onAddToFavorites: (id) => dispatch(Operation.addToFavorites(id)),
  onRemoveFromFavorites: (id) => dispatch(Operation.removeFromFavorites(id)),
});

const enhance = compose(
    withRouter,
    connect(mapStateToProps, mapDispatchToProps)
);

export default enhance(FavoriteButton);
