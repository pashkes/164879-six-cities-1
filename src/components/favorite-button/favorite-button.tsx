import * as React from "react";
import {connect} from "react-redux";
import {withRouter} from "react-router-dom";
import {compose} from "recompose";

import {Operation} from "../../reducer/data/data";
import {getAuthorizationStatus} from "../../reducer/user/selectors";
import Constants from "./../../constants";

interface Props {
  isFavorite: boolean,
  onAddToFavorites: (id: number) => void,
  onRemoveFromFavorites: (id: number) => void,
  id: number,
  prefixClass: string,
  width?: number,
  height?: number,
  isAuthRequire: boolean,
  history: any,
}

export class FavoriteButton extends React.PureComponent<Props, null> {
  handleClick = () => {
    const {
      isFavorite,
      onAddToFavorites,
      id,
      onRemoveFromFavorites,
      isAuthRequire,
      history,
    } = this.props;

    if (isAuthRequire) {
      history.push(Constants.LOGIN_PATH);
      return;
    }
    if (isFavorite) {
      onRemoveFromFavorites(id);
    } else {
      onAddToFavorites(id);
    }
  };

  render() {
    const {isFavorite, prefixClass, width = 18, height = 19} = this.props;

    return (
      <button
        onClick={this.handleClick}
        type="button"
        className={`${prefixClass}__bookmark-button ${isFavorite ? `place-card__bookmark-button--active` : ``} button`}
        aria-pressed={isFavorite ? `true` : `false`}
      >
        <svg className={`place-card__bookmark-icon`} aria-hidden="true" width={width} height={height}>
          <use xlinkHref="#icon-bookmark"/>
        </svg>
        <span className="visually-hidden">{isFavorite ? `In` : `To`} bookmarks</span>
      </button>
    );
  }
}

const mapStateToProps = (state) => ({
  isAuthRequire: getAuthorizationStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  onAddToFavorites: (id) => dispatch(Operation.addToFavorites(id)),
  onRemoveFromFavorites: (id) => dispatch(Operation.removeFromFavorites(id)),
});

const enhance = compose(
    withRouter,
    connect(mapStateToProps, mapDispatchToProps)
);

export default enhance(FavoriteButton);
