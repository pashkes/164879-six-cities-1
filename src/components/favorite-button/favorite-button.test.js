import React from "react";
import {shallow, configure} from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import {FavoriteButton} from "./favorite-button.jsx";

configure({adapter: new Adapter()});

describe(`the button should correctly  is rendered `, () => {

  it(`if the button isn't favorite`, () => {
    const favoriteButton = shallow(<FavoriteButton
      isFavorite={false}
      id={0}
      prefixClass={``}
      isAuthRequire={true}
      history={{}}
      onAddToFavorites={jest.fn()}
      onRemoveFromFavorites={jest.fn()}
    />);
    expect(favoriteButton).toMatchSnapshot();
  });

  it(`if the button is favorite`, () => {
    const favoriteButton = shallow(<FavoriteButton
      isFavorite={true}
      id={0}
      prefixClass={``}
      isAuthRequire={true}
      onAddToFavorites={jest.fn()}
      onRemoveFromFavorites={jest.fn()}
      history={{}}
    />);
    expect(favoriteButton).toMatchSnapshot();
  });
});
