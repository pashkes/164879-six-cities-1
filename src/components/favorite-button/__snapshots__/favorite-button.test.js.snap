// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the button should correctly  is rendered  if the button is favorite 1`] = `
<button
  aria-pressed="true"
  className="__bookmark-button place-card__bookmark-button--active button"
  onClick={[Function]}
  type="button"
>
  <svg
    aria-hidden="true"
    className="place-card__bookmark-icon"
    height={19}
    width={18}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    In
     bookmarks
  </span>
</button>
`;

exports[`the button should correctly  is rendered  if the button isn't favorite 1`] = `
<button
  aria-pressed="false"
  className="__bookmark-button  button"
  onClick={[Function]}
  type="button"
>
  <svg
    aria-hidden="true"
    className="place-card__bookmark-icon"
    height={19}
    width={18}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To
     bookmarks
  </span>
</button>
`;
